src/iosrv/srv.c

```diff
@@ -217,22 +229,98 @@ dss_sched_unit_pop(ABT_pool *pools, ABT_pool *pool)
 	rc = ABT_pool_get_total_size(pools[DSS_POOL_URGENT], &cnt);
 	if (rc != ABT_SUCCESS)
 		return ABT_UNIT_NULL;
-	if (cnt != 0 && rand() % 100 <= dss_first_res_percentage)
+	if (cnt != 0 && fast_rand() % 100 <= dss_first_res_percentage)
 		return unit_pop(pools, DSS_POOL_URGENT, pool);
 
 	/* then pop other pools */
 	rc = ABT_pool_get_total_size(pools[DSS_POOL_REBUILD], &cnt);
 	if (rc != ABT_SUCCESS)
 		return ABT_UNIT_NULL;
 
-	if (cnt == 0 || rand() % 100 >= dss_rebuild_res_percentage)
+	if (cnt == 0 || fast_rand() % 100 >= dss_rebuild_res_percentage)
 		return normal_unit_pop(pools, pool);
 	else
 		return unit_pop(pools, DSS_POOL_REBUILD, pool);
 
 	return ABT_UNIT_NULL;
 }
 
+static struct dss_xstream *
+dss_xstream_get(int stream_id)
+{
+	if (stream_id == DSS_XS_SELF)
+		return dss_get_module_info()->dmi_xstream;
+
+	D_ASSERTF(stream_id >= 0 && stream_id < xstream_data.xd_xs_nr,
+		  "invalid stream id %d (xstream_data.xd_xs_nr %d).\n",
+		  stream_id, xstream_data.xd_xs_nr);
+
+	return xstream_data.xd_xs_ptrs[stream_id];
+}
+
+struct timeout_ult {
+	ABT_thread	tu_thread;
+	uint64_t	tu_expire_time;
+	d_list_t	tu_list;
+};
+
+/* Add the timeout tu to the sorted list */
+static void
+ds_add_expire_list(struct timeout_ult *new_tu)
+{
+	struct dss_xstream *dx;
+	struct timeout_ult *tu;
+
+	dx = dss_xstream_get(DSS_XS_SELF);
+
+	d_list_for_each_entry(tu, &dx->dx_timeout_ult_list, tu_list) {
+		if (tu->tu_expire_time > new_tu->tu_expire_time) {
+			d_list_add_tail(&new_tu->tu_list, &tu->tu_list);
+			return;
+		}
+	}
+
+	d_list_add_tail(&new_tu->tu_list, &dx->dx_timeout_ult_list);
+}
+
+/* Attach the ULT to its own dx */
+int
+dss_sched_timeout(ABT_thread thread, uint64_t timeout)
+{
+	struct timeout_ult *tu;
+
+	D_ALLOC_PTR(tu);
+	if (tu == NULL)
+		return -DER_NOMEM;
+
+	tu->tu_expire_time = d_timeus_secdiff(timeout);
+	tu->tu_thread = thread;
+	ds_add_expire_list(tu);
+
+	ABT_self_suspend();
+	return 0;
+}
+
+static void
+ds_check_expire_list(bool shutdown)
+{
```

Comment:
We could also store the time of the next event and totally touching the list if current_time < time_of_next_event. It would probably save a few millions of instructions overall.

Reply:
indeed. Actually this is already sorted list, so it will break immediately once the current_time < the current expire time.